Assets/MixedRealityToolkit.Staging/UnityAR/Editor/ConfigurationChecker/UnityARConfigurationChecker.cs

```diff
@@ -15,78 +17,108 @@ public class UnityARConfigurationChecker
-                return null;
+                ScriptingUtilities.AppendScriptingDefinitions(BuildTargetGroup.Android, definitions);
+                ScriptingUtilities.AppendScriptingDefinitions(BuildTargetGroup.iOS, definitions);
+                return true;
+            }
+            else
+            {
+                ScriptingUtilities.RemoveScriptingDefinitions(BuildTargetGroup.Android, definitions);
+                ScriptingUtilities.RemoveScriptingDefinitions(BuildTargetGroup.iOS, definitions);
+                return false;
             }
-
-            return files[0];
         }
 
         /// <summary>
-        /// Determines if a file update is needed and returns the replacement.
+        /// Updates the assembly definition to contain the appropriate references based on the Unity
+        /// version and if the project contains the AR Foundation package.
         /// </summary>
-        /// <param name="file">The reference file used to determine the new file.</param>
-        /// <returns>FileInfo object representing the file, or null if a new file is not needed.</returns>
-        private static FileInfo GetNewAsmDefFile(FileInfo file)
+        private static void UpdateAsmDef(bool arFoundationPresent)
         {
-            // Read the file.
-            string fileContents = string.Empty;
-            using (FileStream fs = new FileStream(file.FullName, FileMode.Open, FileAccess.Read))
+            string asmDefFileName = "Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef";
+            FileInfo[] asmDefFiles = FileUtilities.FindFilesInAssets(asmDefFileName);
+
+            if (asmDefFiles.Length == 0)
             {
-                using (StreamReader reader = new StreamReader(fs))
-                {
-                    fileContents = reader.ReadToEnd();
-                }
+                Debug.LogWarning($"Unable to locate file: {asmDefFileName}");
+                return;
+            }
+
+            AssemblyDefinition asmDef = AssemblyDefinition.Load(asmDefFiles[0].FullName);
+            if (asmDef == null)
+            {
+                Debug.LogWarning($"Unable to load file: {asmDefFileName}");
+                return;
+            }
+
+            List<string> references = new List<string>();
+            if (asmDef.References != null)
+            {
+                references.AddRange(asmDef.References);
             }
 
-            // Key values used to determine if a replacement is required.
+            // Assemblies required by Unity 2018 and / or 2019 (and newer)
             const string arFoundationReference = "Unity.XR.ARFoundation";
             const string spatialTrackingReference = "UnityEngine.SpatialTracking";
+            bool changed = false;
 
-            string updateTo = string.Empty;
-#if UNITY_2018
-            if (!fileContents.Contains(arFoundationReference) ||
-                fileContents.Contains(spatialTrackingReference))
+            if (arFoundationPresent)
```

Comment:
> if (arFoundationPresent) [](start = 11, length = 25)

There's a lot of really interesting logic here that's happening due to some relatively complex/history-filled logic around the differences between 2018 and 2019. Can you leave some comments here about the differences/significance of some of these things, so that others can also share in the background as well?

Reply:
It looks like Unity 2018.4.13 added the same behavior as Unity 2019, where missing assembly references are "fine" and the assembly still loads correctly:

![image](https://user-images.githubusercontent.com/3580640/71221330-211e9780-2281-11ea-9056-89283c7416b1.png)

Proposal: Remove this code that manually loads and writes into the assembly definition and instead update our min version to be 2018.4.13 and take advantage of Unity's way of handling this scenario.

Reply:
we have several customers on older versions of 2018.4. We should keep the current loading/writing code until we formally require newer patch versions of 2018.4.

Reply:
Added remarks comment block and additional inline comment.

```
        /// <remarks>
        /// Versions of Unity may have different factorings of components. To address this, the UpdateAsmDef
        /// method conditionally compiles for the version currently in use.
        /// To ensure proper compilation on each Unity version, the following steps are performed:
        /// - Load the Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef file
        /// - If AR Foundation has been installed via the Unity Package Manager, add the appropriate assembly references
        ///   - Unity 2018: Unity.XR.ARFoundation
        ///   - Unity 2019 and newer: Unity.XR.ARFoundation, UnityEngine.SpatialTracking
        /// - If AR Foundatrion has been uninstalled, remove the assembly references. 
        /// - Save the Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef file
        /// This will result in Unity reloading the assembly with the appropriate dependencies.
        /// </remarks>
```

Reply:
>until we formally require newer patch versions of 2018.4

I feel that this could be a reason to do so